src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp

```diff
@@ -82,315 +99,510 @@ enum FourHorsemen
-                uiEventStarterGUID = who->GetGUID();
-                BeginFourHorsemenMovement();
+            if (!_ourMovementFinished || !UpdateVictim())
+                return;
+            events.Update(diff);
 
-                if (!encounterActionAttack)
-                    DoEncounteraction(who, true, false, false);
-            }
-            else if (movementCompleted && movementStarted)
+            while (uint32 eventId = events.ExecuteEvent())
             {
-                if (caster)
-                    me->Attack(who, false);
-                else
-                    BossAI::AttackStart(who);
+                switch (eventId)
+                {
+                    case EVENT_BERSERK:
+                        DoCastAOE(SPELL_BERSERK, true);
+                        break;
+                    case EVENT_MARK:
+                        DoCastAOE(SPELL_THANE_MARK, true);
+                        events.ScheduleEvent(EVENT_MARK, 12 * IN_MILLISECONDS);
+                        break;
+                    case EVENT_METEOR:
+                        if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, 20.0f, true))
+                            DoCast(target, SPELL_METEOR);
+                        events.ScheduleEvent(EVENT_METEOR, 15 * IN_MILLISECONDS);
+                        break;
+                }
             }
+
+            if (me->HasUnitState(UNIT_STATE_CASTING))
+                return;
+            DoMeleeAttackIfReady();
         }
+    };
+
+    CreatureAI* GetAI(Creature* creature) const override
+    {
+        return GetInstanceAI<boss_four_horsemen_thaneAI>(creature);
+    }
+};
 
-        void KilledUnit(Unit* /*victim*/) override
+static const Position ladyPath[3] = { { 2507.94f, -2961.444f, 242.4557f }, { 2488.763f, -2960.007f, 241.2757f }, { 2468.26f, -2947.499f, 241.2753f } };
+class boss_four_horsemen_lady : public CreatureScript
+{
+    public:
+    boss_four_horsemen_lady() : CreatureScript("boss_four_horsemen_lady") { }
+
+    struct boss_four_horsemen_ladyAI : public boss_four_horsemen_baseAI
+    {
+        boss_four_horsemen_ladyAI(Creature* creature) : boss_four_horsemen_baseAI(creature, LADY, ladyPath) { }
+        void BeginFighting() override
         {
-            if (!(rand32() % 5))
-                Talk(SAY_SLAY);
+            events.ScheduleEvent(EVENT_BERSERK, 10 * MINUTE * IN_MILLISECONDS);
+            events.ScheduleEvent(EVENT_MARK, 24 * IN_MILLISECONDS);
+            // @todo actual timing, this is for test
+            events.ScheduleEvent(EVENT_VOIDZONE, urandms(15, 20));
         }
 
-        void JustDied(Unit* /*killer*/) override
+        void _UpdateAI(uint32 diff) override
         {
-            events.Reset();
-            summons.DespawnAll();
+            if (!me->IsInCombat())
+                return;
+            if (!_ourMovementFinished)
+                return;
+            if (me->getThreatManager().isThreatListEmpty())
+                return EnterEvadeMode();
```

Comment:
please use "return;" when the return type is void otherwise it might look like the called function is returning some kind of value.